src/content/profile-data.js

```diff
@@ -527,3 +527,285 @@ export function getTracingMarkers(thread, markers) {
+    for (let stackDepth = depth + 1; stackDepth <= previousDepth; stackDepth++) {
+      // Push on the new information.
+      stackTimingByDepth[stackDepth].start.push(lastSeenStartTimeByDepth[stackDepth]);
+      stackTimingByDepth[stackDepth].end.push(sampleTime);
+      stackTimingByDepth[stackDepth].stack.push(lastSeenStackIndexByDepth[stackDepth]);
+      stackTimingByDepth[stackDepth].length++;
+
+      // Delete that this stack frame has been seen.
+      lastSeenStackIndexByDepth[stackDepth] = undefined;
+      lastSeenStartTimeByDepth[stackDepth] = undefined;
+    }
+  }
+
+  function pushStacks(depth, startingIndex, sampleTime) {
+    let stackIndex = startingIndex;
+    // "Push" onto the stack with new frames
+    for (let parentDepth = depth; parentDepth >= 0; parentDepth--) {
+      if (lastSeenStackIndexByDepth[parentDepth] !== undefined) {
+        break;
+      }
+      lastSeenStackIndexByDepth[parentDepth] = stackIndex;
+      lastSeenStartTimeByDepth[parentDepth] = sampleTime;
+      stackIndex = thread.stackTable.prefix[stackIndex];
+    }
+  }
+
+  // Go through each sample, and push/pop it on the stack to build up
+  // the stackTimingByDepth.
+  let previousDepth = 0;
+  for (let i = 0; i < thread.samples.length; i++) {
+    const stackIndex = thread.samples.stack[i];
+    const sampleTime = thread.samples.time[i];
+    const funcStackIndex = stackIndexToFuncStackIndex[stackIndex];
+    const depth = funcStackTable.depth[funcStackIndex];
+
+    // If the two samples at the top of the stack are different, pop the last stack frame.
+    const depthToPop = lastSeenStackIndexByDepth[depth] === stackIndex ? depth : depth - 1;
+    popStacks(depthToPop, previousDepth, sampleTime);
+    pushStacks(depth, stackIndex, sampleTime);
+    previousDepth = depth;
+  }
+
+  // Pop the remaining stacks
+  const endingTime = thread.samples.time[thread.samples.time.length - 1] + interval;
+  popStacks(-1, previousDepth, endingTime);
+
+  // Collapse platform code into single rows if JS only.
+  if (jsOnly) {
+    collapsePlatformStacks(stackTimingByDepth, thread, funcStackInfo);
+  }
+
+  return stackTimingByDepth;
+}
+
+/**
+ * Given timing information like below, collapse out the platform stacks. In the diagram
+ * "J" represents JavaScript stack frame timing, and "P" Platform stack frame timing.
+ * The timing stack index gets changed to -1 for collapsed platform code.
+ *
+ * JJJJJJJJJJJJJJJJ  --->  JJJJJJJJJJJJJJJJ
+ * PPPPPPPPPPPPPPPP        PPPPPPPPPPPPPPPP
+ *     PPPPPPPPPPPP            JJJJJJJJ
+ *     PPPPPPPP                JJJ  JJJ
+ *     JJJJJJJJ
+ *     JJJ  JJJ
+ *
+ * @param {Object} stackTimingByDepth - Table of stack timings.
+ * @param {Object} thread - The current thread.
+ * @param {Object} funcStackInfo - Info about funcStacks.
+ * @returns {Object} The mutated and collapsed timing information.
+ */
+function collapsePlatformStacks(stackTimingByDepth, thread, funcStackInfo) {
```

Comment:
Hmm... my thought was testing the exported interface, not the internal components. There is coverage for it because I explicitly test a full profile, and a JS-only view that needs the platform stacks collapsed.